### PR TITLE
Cleanup old dismissed messages, test InnoDB

### PR DESCRIPTION
**Pull request by patrickxb**
The cleanup application will delete old dismissed messages in the gregor database and their related rows in other tables.

Also, tested innodb file size changes with delete operations.  See doc/mysql.md for results.

r? @maxtaco

### CODE REVIEW

File: doc/mysql.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+2. `messages.ibd` after is 236MB.
+3. `DELETE FROM messages;` to remove all the rows.
+4. `messages.ibd` after is 236MB (as expected).
+5. Added 100k rows, `messages.ibd` still 236MB.
+6. Added 1M rows (total), `messages.ibd` still 236MB.
+7. Added 1.1M rows (total), `messages.ibd` now 252MB.
+
+Conclusion: while the InnoDB file size doesn't shrink
+after deleting rows, the space that was used by deleted 
+rows is reused for new rows.
```
Comment (maxtaco) [inline doc/mysql.md]: Which version of mysql and which configuration parameters did you use?

**Comment by patrickxb** [comment]
Also, note this followup ticket:  https://keybase.atlassian.net/browse/CORE-2935

File: storage/sql_schema.go [hunk 2]
```diff
@@ -35,14 +35,16 @@ var schema = []string{
...
 		PRIMARY KEY(uid, msgid, ntime)
 	)`,
 
 	`CREATE TABLE dismissals_by_id (
 		uid   CHAR(32) NOT NULL,
 		msgid CHAR(32) NOT NULL,
 		dmsgid CHAR(32) NOT NULL, -- "the message IDs to dismiss",
-		FOREIGN KEY(uid, msgid) REFERENCES messages (uid, msgid),
+		FOREIGN KEY(uid, msgid) REFERENCES messages (uid, msgid) ON DELETE CASCADE,
+		FOREIGN KEY(uid, dmsgid) REFERENCES messages (uid, msgid) ON DELETE CASCADE,
```
Comment (maxtaco) [inline storage/sql_schema.go]: is sql still ok with this?

File: cleanup/cleanup.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+func main() {
+	var interval int
+	var help bool
+	var dsn string
+
+	flag.IntVar(&interval, "interval", 30, "clean messages older than interval `days`")
+	flag.BoolVar(&help, "h", false, "show help")
+	flag.BoolVar(&help, "help", false, "show help")
+	flag.StringVar(&dsn, "dsn", "gregor:@/gregor_test", "mysql dsn for gregor database")
```
Comment (maxtaco) [inline cleanup/cleanup.go]: can this script work for sqlite in any meaningful way?

File: doc/mysql.md (hunk 1 above)
Comment (patrickxb) [inline doc/mysql.md]: 5.7.10 with default parameters on OS X.

File: cleanup/cleanup.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	db, err := sql.Open("mysql", dsn)
+	if err != nil {
+		log.Fatalf("error opening database connection to %s: %s", dsn, err)
+	}
+	return db
+}
+
+func cleanup(db *sql.DB, interval int) {
+	for i := 0x00; i <= 0xff; i++ {
+		partition(db, fmt.Sprintf("%02x", i), interval)
```
Comment (maxtaco) [inline cleanup/cleanup.go]: Can you have parameters for partition size (in fraction of the DB) and also a sleep between partition attempts?  Also, how about daemon mode, to be running persistently?

File: doc/mysql.md (hunk 1 above)
Comment (maxtaco) [inline doc/mysql.md]: I think Aurora runs 5.5 or 5.6, no?

File: storage/sql_schema.go (hunk 2 above)
Comment (patrickxb) [inline storage/sql_schema.go]: sqlite tests pass.

File: cleanup/cleanup.go (hunk 3 above)
Comment (patrickxb) [inline cleanup/cleanup.go]: I believe it works the same way:

https://www.sqlite.org/foreignkeys.html#fk_actions

File: cleanup/cleanup.go (hunk 4 above)
Comment (patrickxb) [inline cleanup/cleanup.go]: Sure.

File: storage/sql_schema.go (hunk 2 above)
Comment (patrickxb) [inline storage/sql_schema.go]: See this link:  https://www.sqlite.org/foreignkeys.html#fk_actions

File: doc/mysql.md (hunk 1 above)
Comment (patrickxb) [inline doc/mysql.md]: 5.6.

I'll test with 5.6.  Would be hard to test with aurora itself as no access to the actual files, but I'll test with 5.6 on a VM.

Comment (patrickxb) [inline doc/mysql.md]: Same results on 5.6.

**Comment by patrickxb** [comment]
PTAL

File: cleanup/cleanup.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	tx, err := db.Begin()
+	if err != nil {
+		log.Fatalf("error starting transaction: %s", err)
+	}
+	defer tx.Commit()
+
+	// Note that this query relies on the following foreign key reference
+	// FOREIGN KEY(uid, msgid) REFERENCES messages (uid, msgid) ON DELETE CASCADE
+	// on child tables of messages.
+	res, err := tx.Exec("DELETE FROM messages USING messages LEFT JOIN items ON (messages.uid=items.uid AND messages.msgid=items.msgid) WHERE items.uid LIKE ? AND items.dtime < DATE_SUB(NOW(), INTERVAL ? DAY)", prefix+"%", interval)
```
Comment (maxtaco) [inline cleanup/cleanup.go]: can you format this nicer?

File: doc/mysql.md (hunk 1 above)
Comment (maxtaco) [inline doc/mysql.md]: can you add these notes into the mysql.md file?

Comment (patrickxb) [inline doc/mysql.md]: I did already

File: cleanup/cleanup.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func partitionFormat() {
+	maxHex := fmt.Sprintf("%x", partitions-1)
+	for _, x := range maxHex {
+		if x != 'f' {
+			log.Fatalf("number of partitions: %d, must be a power of 16", partitions)
+		}
+	}
+	format = fmt.Sprintf("%%0%dx", len(maxHex))
```
Comment (maxtaco) [inline cleanup/cleanup.go]: this is mad tricky, code to generate a format string, might need a comment or two....

File: cleanup/cleanup.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	select {
+	case x := <-c:
+		log.Printf("caught signal %v", x)
+		close(done)
+	}
+}
+
+func partitionFormat() {
+	maxHex := fmt.Sprintf("%x", partitions-1)
+	for _, x := range maxHex {
```
Comment (maxtaco) [inline cleanup/cleanup.go]: this loop too can use a comment, etc...

**Comment by maxtaco** [comment]
👍 after comments, docs, and formatting!

File: doc/mysql.md (hunk 1 above)
Comment (maxtaco) [inline doc/mysql.md]: oh, don't see the notes ?

Comment (patrickxb) [inline doc/mysql.md]: lines 18-20.

Comment (maxtaco) [inline doc/mysql.md]: i see now, github was hiding them from me

**Pull request merged**